test/message/_message.cxx

```diff
@@ -182,13 +194,164 @@ auto EmptyMessage(string const& transport, string const& _address) -> void
     push.Bind(address);
     pull.Connect(address);
 
-    auto outMsg(push.NewMessage());
-    ASSERT_EQ(outMsg->GetData(), nullptr);
-    ASSERT_EQ(push.Send(outMsg), 0);
+    {
+        auto outMsg(push.NewMessage());
+        ASSERT_EQ(outMsg->GetData(), nullptr);
+        ASSERT_EQ(push.Send(outMsg), 0);
+
+        auto inMsg(pull.NewMessage());
+        ASSERT_EQ(pull.Receive(inMsg), 0);
+        ASSERT_EQ(inMsg->GetData(), nullptr);
+    }
 
-    auto inMsg(pull.NewMessage());
-    ASSERT_EQ(pull.Receive(inMsg), 0);
-    ASSERT_EQ(inMsg->GetData(), nullptr);
+    {
+        auto outMsg(push.NewMessage(0));
+        ASSERT_EQ(outMsg->GetSize(), 0);
+
+        size_t const size100{100};
+        outMsg->Rebuild(size100);
+        ASSERT_EQ(outMsg->GetSize(), size100);
+
+        outMsg->Rebuild(0);
+        ASSERT_EQ(outMsg->GetSize(), 0);
+
+        auto msgCopy(push.NewMessage());
+        msgCopy->Copy(*outMsg);
+        ASSERT_EQ(msgCopy->GetSize(), 0);
+        ASSERT_EQ(push.Send(outMsg), 0);
+        ASSERT_EQ(push.Send(msgCopy), 0);
+
+        auto inMsg(pull.NewMessage());
+        auto inMsg2(pull.NewMessage());
+        ASSERT_EQ(pull.Receive(inMsg), 0);
+        ASSERT_EQ(pull.Receive(inMsg2), 0);
+        ASSERT_EQ(inMsg->GetSize(), 0);
+        ASSERT_EQ(inMsg2->GetSize(), 0);
+    }
+}
+
+// The "zero copy" property of the Copy() method is an implementation detail and is not guaranteed.
+// Currently it holds true for the shmem (across devices) and for zeromq (within same device) transports.
+auto ZeroCopy(const string& transport) -> void
+{
+    ProgOptions config;
+    config.SetProperty<string>("session", tools::Uuid());
+    auto factory(TransportFactory::CreateTransportFactory(transport, tools::Uuid(), &config));
+
+    unique_ptr<string> str(make_unique<string>("asdf"));
+    const size_t size = 2;
+    MessagePtr original(factory->CreateMessage(size));
+    memcpy(original->GetData(), "AB", size);
+    {
+        MessagePtr copy(factory->CreateMessage());
+        copy->Copy(*original);
+        EXPECT_EQ(original->GetSize(), copy->GetSize());
+        EXPECT_EQ(original->GetData(), copy->GetData());
+        EXPECT_EQ(static_cast<const shmem::Message&>(*original).GetRefCount(), 2);
+        EXPECT_EQ(static_cast<const shmem::Message&>(*copy).GetRefCount(), 2);
```

Comment:
```suggestion
    if ("shmem" == transport) {
        EXPECT_EQ(static_cast<const shmem::Message&>(*original).GetRefCount(), 2);
        EXPECT_EQ(static_cast<const shmem::Message&>(*copy).GetRefCount(), 2);
    }
```

Reply:
alternatively, remove the transport function argument or assert it to be "shmem".